drivers/sdcard.py Fix failure with shared SPI bus

peterhinch opened the pull request:
If the SPI bus is shared with another device and that device is accessed concurrently with the SD card, the driver fails with `EIO` or timeouts. This patch writes >100 bus cycles with `MOSI` high prior to any access.

The fault may be demonstrated (with my SD card) with the following script:
```python
from machine import SPI, Pin
import sdcard
import os
spi = SPI(2)  # 2 MOSI Y8 MISO Y7 SCK Y6
sdcs = Pin('Y4', Pin.OUT, value=1)  # SD card CS
sd = sdcard.SDCard(spi, sdcs)
vfs = os.VfsFat(sd)
os.mount(vfs, '/fc')
os.listdir('/fc')
spi.write(b'\x00'*50)  # Write to notional device
with open('/fc/yellow.mp3', 'rb') as f:  # File must exist!
    f.read(20)
```

peterhinch committed:
Clock card with >100 cycles before access

In case bus was used between accesses.

tve commented:
I haven't used SDcards on microcontrollers, so pardon the ignorance: is this 100-cycles-MOSI-high thing something commonly done for SDcards? It feels like a sledge hammer approach :-O.

Also, you seem to experience the issue on a pyboard? Or other ports as well? The patch changes the driver for all ports...

peterhinch commented:
The current driver's initialisation code issues 100+ cycles with MOSI high and CS deselected. This is evidently necessary for correct initialisation of the card. In my testing, if the bus is used for other purposes, this is also required before every access.

This requirement is a characteristic of SD cards and is not host-specific; the need for it applies to all ports. Also to both soft and hard SPI.

tve commented:
Sharing an SPI bus with an SDcard seems tricky according to some googling around... However, the answer to this question may have to do with your issue: https://electronics.stackexchange.com/questions/68551/does-an-sd-card-in-spi-mode-respect-chip-select-slave-select-seems-to-be-resett

peterhinch commented:
Thanks for the useful link. I tested this approach, writing a single byte rather than 16 and it works. With the inevitable proviso "on my SD card".

Prior experience with this driver has left me with a sense that driving SD cards is something of a black art. We have already implemented changes in response to failures with certain brands of card.

My reason for using multiple bytes was because the initialisation code already does this, with an explicit comment about sending >100 edges with CS false. With block sizes of 512KiB this represents a time overhead of ~3%. But it is inelegant if  unnecessary.

I would welcome guidance from anyone with relevant knowledge.

peterhinch commented:
I replicated the result in my OP on hardware where the SPI bus is not shared. Sharing is simulated by sending a single 0 byte when CS is False. This is enough to trigger failure unless a single `b'\xff'` is sent at the start of `.readblocks`. See [this post](https://github.com/adafruit/Adafruit_CircuitPython_VS1053/issues/11#issuecomment-626296352) for script and context.

@tve This seems to be a different effect from that you referenced, which is addressed by the official driver. This sends `b'\xff'` at the end of transfers which should ensure bus release. In this issue the fault is at the start of a transfer: prior bus usage prevents a subsequent transfer.

Comments or explanations gratefully received :)

peterhinch committed:
Send a single byte of 0xff only

This works in my testing.

peterhinch commented:
This has languished for six months now. Sending a single byte with CS False seems benign and does fix a bug. Thoughts or comments?

dpgeorge commented:
This looks ok to me.  As you say it's effectively a no-op on the SPI bus for standard SPI devices.  I guess sending 0xff with CS deasserted will reset some internal SPI state on the SD card, which is maybe why it works for the stack-overflow link above.

dpgeorge commented:
I tried to test this on my hardware, but even if I do `spi.write(b'\x00')` while in between SD card calls, the subsequent call to `sdcard.readblocks(0, buf)` still succeeds.  I guess this fix is only needed for some SD cards?

peterhinch commented:
If you can't replicate the fault then it must be card specific as in my testing it was entirely consistent. The card was a Kingston 16GB class 10 SDHC device. There was an interesting discussion with Adafruit [here](https://github.com/adafruit/Adafruit_CircuitPython_VS1053/issues/11#issuecomment-626185472) but they could not enlighten me as to why this fix worked.

The fix has the merit of harmlessness :)

dpgeorge commented:
Here's my test script that runs on a PYBv1.x with an SD card in the built-in SD-card slot:
```python
import os
from machine import Pin, SoftSPI
from sdcard import SDCard

spi = SoftSPI(sck=Pin('SD_CK'), mosi=Pin('SD_CMD'), miso=Pin('SD_D0'))
cs = Pin('SD_D3')
sd = SDCard(spi, cs)

for _ in range(100):
    buf = bytearray(512)
    spi.write(os.urandom(4))
    sd.readblocks(0, buf)
    if buf[:12] != b'\xeb\xfe\x90MSDOS5.0\x00':
        print("fail", buf[:12])
```

peterhinch commented:
That fails here with repeated
```
fail bytearray(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
```
However it fails both with and without my change.
Environment:
```
MicroPython v1.14-67-gcf6a01588-dirty on 2021-02-17; PYBv1.1 with STM32F405RG
```
I'll look at this later to figure out what's happening when I have some time.

dpgeorge commented:
> That fails here with repeated ...

Make sure your SD card is formatted with FAT!  See what `readblocks(0, buf)` gives without any errors.

peterhinch commented:
I had planned to do just that, but had to rush off to get vaccinated ;-) 

Unfortunately I'm getting no joy. The SD card has a filesystem and `gparted` on my PC reports it as FAT32 (screenshot below). Here is the outcome when I try to read sector 0 (with unmodified sdcard.py):
```python
MPY: sync filesystems
MPY: soft reboot
MicroPython v1.14-67-gcf6a01588-dirty on 2021-02-17; PYBv1.1 with STM32F405RG
Type "help()" for more information.
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== import os
=== from machine import Pin, SoftSPI
=== from sdcard import SDCard
=== 
=== spi = SoftSPI(sck=Pin('SD_CK'), mosi=Pin('SD_CMD'), miso=Pin('SD_D0'))
=== cs = Pin('SD_D3')
=== sd = SDCard(spi, cs)
=== 
=== buf = bytearray(512)
=== sd.readblocks(0, buf)
=== print(buf[:12])
=== 
bytearray(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
```
![image](https://user-images.githubusercontent.com/8437477/115430169-086b1c80-a1fc-11eb-957b-8c26ab1d1d1a.png)

robert-hh commented:
I'm a little bit confused about that unallocated space. What I do is to erase the partition and recreate it, such that there is no such unallocated space, especially not at the lower end of the card memory.

dpgeorge commented:
> I had planned to do just that, but had to rush off to get vaccinated ;-)

Fantastic!

> Unfortunately I'm getting no joy. The SD card has a filesystem

You can try a different block, it doesn't have to be block 0.  Eg `sd.readblocks(256, buf)`.  

Can you make your original test fail on PYBv1.x with the built-in SD card reader?

peterhinch commented:
It emerged that the filesystem on the Kingston SD card was corrupt. I deleted the partition, reformatted FAT32 and restored the contents. I still don't see your `b'\xeb\xfe\x90MSDOS5.0\x00'` contents, but that isn't really significant: what we're looking for are exceptions thrown. The outcome of testing your script on the Pyboard 1.1 was as follows. I performed a power cycle between each test and **boot.py** disables MSC mode. I tested on two samples each of Sandisk and Kingston.

1. Sandisk cards threw no exceptions on unmodified **sdcard.py**. Script ran to completion.
2. Kingston cards with unmodified **sdcard.py** threw `OSError: [Errno 5] EIO` before printing any output.
3. On Kingston with my variant of **sdcard.py** the script ran to completion.
4. On Kingston I removed the `spi.write` from the script. Using standard **sdcard.py** it ran to completion.

The second Kingston card also had a corrupt filesystem, fixed prior to testing as aboe. Neither Sandisk sample showed filesystem corruption even though the first had been heavily used for development.

Conclusions:
 1. Sandisk samples did not exhibit the fault.
 2. Kingston samples did exhibit it. The fault appears to be as I described, can be demonstrated with your script, and can be fixed with this patch.

However it is specific to these Kingston cards and there is evidently something odd about them given the initial corruption. This despite the fact that I've been happily using them in normal development. Kingston is a good brand, and I only buy from reputable distributors, but it is possible they are clones.

To add to the confusion I also tested a Samsung 32GB card. The filesystem was clean according to gparted. The script with standard **sdcard.py** couldn't get past initialisation with
```python
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
  File "/flash/sdcard.py", line 54, in __init__
  File "/flash/sdcard.py", line 87, in init_card
  File "/flash/sdcard.py", line 135, in init_card_v2
OSError: timeout waiting for v2 card
```

Conclusion: SD cards are horrible :) I don't know where we go from here.

peterhinch commented:
@dpgeorge In an effort to resolve this I raised [this forum request](https://forum.micropython.org/viewtopic.php?f=6&t=10366&p=57385#p57385) to try to establish whether this is widespread or restricted to my Kingston instances.

In your test script I removed the check on contents as in my testing this always fails to match. An actual failure produces an **OSError**.

I'll report back in due course.

peterhinch commented:
Dave Hylands has confirmed my observations on older Kingston cards. A brand new Kingston device passes the test.

davehylands commented:
Not sure if it was "new" related or size related. I know in the past micropython has had issues with larger cards.

I also noticed that the larger card shows up as "W95 FAT (LBA)" under Linux and the smaller card didn't have the (LBA).

GC-RnD commented:
Peter

I wish somone would validate your modifacations to the sdcard.py driver.
I can confrim that this fix worked for me using a 2GB and 8GB SD Card (amazon crap card),
ILI9341 (display driver), and (xpt2046 touch driver).

Orginal I had a 2GB sd that had no issues. 
When I purchase new cards from amazon is when I got
corrupted read, write and mounting issues.

by adding...
# For shared bus operation
self.spi.write(b"\xff")

under...
line 229  def readblocks(self, block_num, buf):
and...
line 258  def writeblocks(self, block_num, buf):

fixed all issues !!!

PETER...
I would like to use the machine_sdcard.c driver
but I am not confident where or how to make the modifications.
Would you be able to add the fix to this file?

The time I have wasted trying to get these 3 drivers
to coexist on a shared SPI bus is insidious !!!

So now my project uses...
a modified sdcard.py (but would like the machine_sdcard.c)
a modified modILI9341.c from lv_micropython
a modified modxpt2046.c from lv_micropython

And for the curious...
I never got the lv_micropython standard drivers to work.
ili9XXX.py  (display driver)
xpt2046.py  (touch driver)

For a new maker starting out looking to use a SHARED SPI
would absolutely be unnecessarily frustrated and discouraged.

Damien George of micropython and
Amir Gonnen of lv_micropython

REALLY NEED TO ADDRESS THIS ISSUE !!!!!!

peterhinch commented:
I am puzzled that this hasn't been merged. Sending an extra byte with CS `False` is evidently innocuous, and does fix a problem with some SD cards.

jimmo commented:
@peterhinch Makes sense to me. I think we should merge it.

Because sdcard.py has moved to micropython-lib in the meantime, I have re-opened this PR there. https://github.com/micropython/micropython-lib/pull/574

dpgeorge commented:
Sorry that this got neglected.  It's now been merged over at `micropython-lib. (see above link).
